File: Samples/VB.NET/AvocadoSmoothieVBSample/AvocadoSmoothieVBSample/FrmMain.vb
Imports AvocadoSmoothie.Barista
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms

Public Class FrmMain

    Private Shared ReadOnly BoundaryModes As String() = {"Symmetric", "Adaptive", "Replicate", "ZeroPad"}

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        ' Ensure one method is selected by default
        If Not chkMiddleMedian.Checked AndAlso Not chkAllMedian.Checked Then
            chkMiddleMedian.Checked = True
        End If

        ' Remove 32K limit so large outputs can be displayed
        txtInit.MaxLength = 0
        txtInit.Multiline = True
        txtInit.WordWrap = True

        txtRefined.MaxLength = 0
        txtRefined.Multiline = True
        txtRefined.WordWrap = True

        cbxBoundaryMode.Items.Clear()
        cbxBoundaryMode.Items.AddRange(BoundaryModes)
        If cbxBoundaryMode.Items.Count > 0 Then
            cbxBoundaryMode.SelectedIndex = 0
        End If

        UpdateParameterControls()
        UpdateStatus("Ready")
    End Sub

    ' Enable/disable controls based on method selection
    Private Sub UpdateParameterControls()
        ' BoundaryMode : Enable only if AllMedian is checked
        cbxBoundaryMode.Enabled = chkAllMedian.Checked
        lblBoundaryMode.Enabled = chkAllMedian.Checked

        ' BorderCount : Enable only if MiddleMedian is checked
        updBorderCnt.Enabled = chkMiddleMedian.Checked
        lblBorderCnt.Enabled = chkMiddleMedian.Checked
    End Sub

    Private Sub chkAllMedian_CheckedChanged(sender As Object, e As EventArgs) Handles chkAllMedian.CheckedChanged
        UpdateParameterControls()
    End Sub

    Private Sub chkMiddleMedian_CheckedChanged(sender As Object, e As EventArgs) Handles chkMiddleMedian.CheckedChanged
        UpdateParameterControls()
    End Sub

    ' Helper to get the selected boundary mode (for smoothing / export)
    Private Function GetSelectedBoundaryMode() As SignatureMedian.BoundaryMode
        Select Case cbxBoundaryMode.SelectedItem?.ToString()
            Case "Symmetric" : Return SignatureMedian.BoundaryMode.Symmetric
            Case "Adaptive" : Return SignatureMedian.BoundaryMode.Adaptive
            Case "Replicate" : Return SignatureMedian.BoundaryMode.Replicate
            Case "ZeroPad" : Return SignatureMedian.BoundaryMode.ZeroPad
            Case Else : Return SignatureMedian.BoundaryMode.Symmetric
        End Select
    End Function

    Private Async Sub btnStart_Click(sender As Object, e As EventArgs) Handles btnStart.Click
        Try
            SetBusy(True)

            Dim data = ParseInputData(txtInit.Text)
            If data Is Nothing OrElse data.Count = 0 Then
                MessageBox.Show(Me, "Please provide at least one numeric value in the Initial Data box.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information)
                UpdateStatus("No data to process")
                Return
            End If

            Dim radius As Integer = DecimalToIntSafe(updRadius.Value) ' R
            Dim borderCount As Integer = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value))
            If radius <= 0 Then
                MessageBox.Show(Me, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                UpdateStatus("Invalid radius")
                Return
            End If

            Dim windowWidth As Integer = (2 * radius) + 1 ' odd window width W
            Dim runAll As Boolean = chkAllMedian.Checked
            Dim runMiddle As Boolean = chkMiddleMedian.Checked

            If Not runAll AndAlso Not runMiddle Then
                MessageBox.Show(Me, "Please select at least one method: All Median and / or Middle Median.", "Selection Required", MessageBoxButtons.OK, MessageBoxIcon.Information)
                UpdateStatus("No method selected")
                Return
            End If

            ' Validation (same rule used in exports): 2 × B must be strictly less than W
            If runMiddle AndAlso (2 * borderCount) >= windowWidth Then
                Dim resp = MessageBox.Show(Me,
                "For Middle Median, 2 × Border Count must be strictly less than Kernel Width (W = 2R + 1)." & vbCrLf & vbCrLf & "Continue anyway?",
                "Parameter Warning",
                MessageBoxButtons.YesNo,
                MessageBoxIcon.Warning)
                If resp = DialogResult.No Then
                    UpdateStatus("Smoothing canceled by user")
                    Return
                End If
            End If

            ' Clamp Middle Median border count so 2B < (2R + 1) => B <= R - 1
            Dim middleBorder As Integer = borderCount
            Dim maxValidB As Integer = Math.Max(0, radius - 1)
            If runMiddle AndAlso middleBorder > maxValidB Then
                middleBorder = maxValidB
                UpdateStatus(String.Format("Middle Median: Border Count clamped to {0} (R={1})", middleBorder, radius))
            End If

            UpdateStatus(String.Format("Smoothing started (R={0}, W={1})...", radius, windowWidth))
            txtRefined.Clear()

            Dim n = data.Count
            Dim sb As New StringBuilder()

            ' Run All Median if requested
            If runAll Then
                Dim progressAll As IProgress(Of Integer) = New Progress(Of Integer)(
                Sub(vp)
                    Dim percent As Integer = If(n > 0, Math.Min(100, Math.Max(0, CInt(Math.Round((vp / CDbl(n)) * 100)))), 0)
                    UpdateStatus(String.Format("AllMedian progress: {0}%", percent))
                End Sub)

                Dim boundaryMode = GetSelectedBoundaryMode()

                Dim allResult = Await Task.Run(Function()
                                                   Return SignatureMedian.ComputeMedians(
                                                   input:=data.ToList(),
                                                   useMiddle:=False,
                                                   kernelWidth:=windowWidth,
                                                   borderCount:=borderCount,
                                                   progress:=progressAll,
                                                   boundaryMode:=boundaryMode)
                                               End Function)

                sb.AppendLine("[All Median]")
                For Each v In allResult
                    sb.AppendLine(v.ToString("G17", CultureInfo.InvariantCulture))
                Next

                If runMiddle Then
                    sb.AppendLine()
                End If
            End If

            ' Run Middle Median if requested
            If runMiddle Then
                Dim progressMiddle As IProgress(Of Integer) = New Progress(Of Integer)(
                Sub(vp)
                    Dim percent As Integer = If(n > 0, Math.Min(100, Math.Max(0, CInt(Math.Round((vp / CDbl(n)) * 100)))), 0)
                    UpdateStatus(String.Format("MiddleMedian progress: {0}%", percent))
                End Sub)

                Dim middleResult = Await Task.Run(Function()
                                                      Return SignatureMedian.ComputeMedians(
                                                      input:=data.ToList(),
                                                      useMiddle:=True,
                                                      kernelWidth:=windowWidth,    ' API expects W (odd window width)
                                                      borderCount:=middleBorder,   ' clamped if needed
                                                      progress:=progressMiddle)
                                                  End Function)

                sb.AppendLine("[Middle Median]")
                For Each v In middleResult
                    sb.AppendLine(v.ToString("G17", CultureInfo.InvariantCulture))
                Next
            End If

            txtRefined.Text = sb.ToString().TrimEnd(ControlChars.Cr, ControlChars.Lf)
            UpdateStatus("Smoothing completed")
        Catch ex As ArgumentOutOfRangeException
            MessageBox.Show(Me, ex.Message, "Out of Range", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Out of range")
        Catch ex As ArgumentException
            MessageBox.Show(Me, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Argument error")
        Catch ex As InvalidOperationException
            MessageBox.Show(Me, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Operation error")
        Catch ex As AggregateException
            Dim flat = ex.Flatten()
            Dim inner = flat.InnerExceptions.FirstOrDefault()
            If TypeOf inner Is InvalidOperationException Then
                MessageBox.Show(Me, inner.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
                UpdateStatus("Operation error")
            Else
                MessageBox.Show(Me, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
                UpdateStatus("Operation error")
            End If
        Catch ex As Exception
            ' Align unexpected errors with "Operation Error" for btnStart
            MessageBox.Show(Me, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Operation error")
        Finally
            SetBusy(False)
        End Try
    End Sub

    ' Hook this to btnExport.Click in the designer
    Private Async Sub btnExcelExport_Click(sender As Object, e As EventArgs) Handles btnExcelExport.Click
        Dim cts As CancellationTokenSource = Nothing
        Try
            SetBusy(True)

            Dim data = ParseInputData(txtInit.Text)
            If data Is Nothing OrElse data.Count = 0 Then
                MessageBox.Show(Me, "Please provide at least one numeric value in the Initial Data box before exporting.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information)
                UpdateStatus("No data to export")
                Return
            End If

            Dim radius As Integer = DecimalToIntSafe(updRadius.Value)   ' R
            Dim borderCount As Integer = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value))
            If radius <= 0 Then
                MessageBox.Show(Me, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                UpdateStatus("Invalid radius")
                Return
            End If

            Dim windowWidth As Integer = (2 * radius) + 1 ' W = 2R + 1
            If chkMiddleMedian.Checked AndAlso (2 * borderCount) >= windowWidth Then
                Dim resp = MessageBox.Show(Me,
                                           "For Middle Median, 2 × Border Count must be strictly less than Kernel Width (W = 2R + 1)." & vbCrLf & vbCrLf & "Continue anyway?",
                                           "Parameter Warning",
                                           MessageBoxButtons.YesNo,
                                           MessageBoxIcon.Warning)
                If resp = DialogResult.No Then
                    UpdateStatus("Export canceled by user")
                    Return
                End If
            End If

            Dim progress As IProgress(Of Integer) = New Progress(Of Integer)(
                Sub(v)
                    UpdateStatus(String.Format("Excel export: {0}%", v))
                End Sub)

            cts = New CancellationTokenSource()

            ' Build and sanitize dataset title to satisfy Excel worksheet name constraints
            Dim rawTitle = BuildDatasetTitle()
            Dim safeTitle = SanitizeWorksheetName(rawTitle)
            If Not String.Equals(rawTitle, safeTitle, StringComparison.Ordinal) Then
                UpdateStatus(String.Format("Worksheet name adjusted to: {0}", safeTitle))
            End If

            Dim ticket As New ExcelOrderTicket() With {
                .InitialData = data,
                .DatasetTitle = safeTitle,   ' Excel worksheet name; sanitized
                .KernelRadius = radius,      ' ticket expects R (service computes W internally)
                .BorderCount = borderCount,
                .Progress = progress,
                .BoundaryMode = GetSelectedBoundaryMode(),
                .CancellationToken = cts.Token
            }

            UpdateStatus(String.Format("Launching Excel export (R={0}, W={1})...", radius, windowWidth))
            Await ExcelBrewService.ExcelCustomOrder(ticket)

            UpdateStatus("Excel export completed")
        Catch ex As ArgumentException
            MessageBox.Show(Me, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Argument error")
        Catch ex As InvalidOperationException
            MessageBox.Show(Me, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Operation error")
        Catch ex As Runtime.InteropServices.COMException
            MessageBox.Show(Me, ex.Message, "Excel Interop Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Interop error")
        Catch ex As OperationCanceledException
            UpdateStatus("Export canceled")
        Catch ex As Exception
            MessageBox.Show(Me, ex.Message, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Error")
        Finally
            If cts IsNot Nothing Then cts.Dispose()
            SetBusy(False)
        End Try
    End Sub

    Private Async Sub btnCSVExport_Click(sender As Object, e As EventArgs) Handles btnCSVExport.Click
        Dim cts As CancellationTokenSource = Nothing
        Try
            SetBusy(True)

            Dim data = ParseInputData(txtInit.Text)
            If data Is Nothing OrElse data.Count = 0 Then
                MessageBox.Show(Me, "Please provide at least one numeric value in the Initial Data box before exporting.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information)
                UpdateStatus("No data to export")
                Return
            End If

            Dim radius As Integer = DecimalToIntSafe(updRadius.Value) ' R
            Dim borderCount As Integer = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value))
            If radius <= 0 Then
                MessageBox.Show(Me, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                UpdateStatus("Invalid radius")
                Return
            End If

            Dim windowWidth As Integer = (2 * radius) + 1 ' W
            If chkMiddleMedian.Checked AndAlso (2 * borderCount) >= windowWidth Then
                Dim resp = MessageBox.Show(Me,
                                           "For Middle Median, 2 × Border Count must be strictly less than Kernel Width (W = 2R + 1)." & vbCrLf & vbCrLf & "Continue anyway?",
                                           "Parameter Warning",
                                           MessageBoxButtons.YesNo,
                                           MessageBoxIcon.Warning)
                If resp = DialogResult.No Then
                    UpdateStatus("CSV export canceled by user")
                    Return
                End If
            End If

            ' Ask user where to save CSV
            Dim title = BuildDatasetTitle()
            Dim sfd As New SaveFileDialog() With {
                .Title = "Save CSV",
                .Filter = "CSV files (*.csv)|*.csv|All files (*.*)|*.*",
                .FileName = SanitizeFileName(title) & ".csv",
                .AddExtension = True,
                .OverwritePrompt = True,
                .SupportMultiDottedExtensions = True
            }

            If sfd.ShowDialog(Me) <> DialogResult.OK Then
                UpdateStatus("CSV export canceled")
                Return
            End If

            Dim progress As IProgress(Of Integer) = New Progress(Of Integer)(
                Sub(v)
                    UpdateStatus(String.Format("CSV export: {0}%", v))
                End Sub)

            cts = New CancellationTokenSource()

            Dim ticket As New CsvOrderTicket() With {
                .InitialData = data,
                .DatasetTitle = title,
                .KernelRadius = radius,   ' ticket expects R; service computes W internally
                .BorderCount = borderCount,
                .BasePath = sfd.FileName,
                .Progress = progress,
                .BoundaryMode = GetSelectedBoundaryMode(),
                .CancellationToken = cts.Token
            }

            UpdateStatus(String.Format("Launching CSV export (R={0}, W={1})...", radius, windowWidth))
            Await CsvBrewService.ExportCsvAsync(ticket)

            UpdateStatus("CSV export completed")
        Catch ex As ArgumentException
            MessageBox.Show(Me, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Argument error")
        Catch ex As DirectoryNotFoundException
            MessageBox.Show(Me, ex.Message, "Directory Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Directory error")
        Catch ex As InvalidOperationException
            MessageBox.Show(Me, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Operation error")
        Catch ex As OperationCanceledException
            UpdateStatus("CSV export canceled")
        Catch ex As Exception
            MessageBox.Show(Me, ex.Message, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            UpdateStatus("Error")
        Finally
            If cts IsNot Nothing Then cts.Dispose()
            SetBusy(False)
        End Try
    End Sub

    Private Shared Function DecimalToIntSafe(value As Decimal) As Integer
        If value > Integer.MaxValue Then Return Integer.MaxValue
        If value < Integer.MinValue Then Return Integer.MinValue
        Return CInt(value)
    End Function

    Private Function ParseInputData(text As String) As IList(Of Double)
        If String.IsNullOrWhiteSpace(text) Then
            Return New List(Of Double)()
        End If

        Dim tokens = text.Split(New Char() {ControlChars.Cr, ControlChars.Lf, ","c, ";"c, ControlChars.Tab, " "c}, StringSplitOptions.RemoveEmptyEntries)
        Dim list As New List(Of Double)(tokens.Length)
        For Each token In tokens
            Dim value As Double
            If Double.TryParse(token, NumberStyles.Float Or NumberStyles.AllowThousands, CultureInfo.InvariantCulture, value) OrElse
               Double.TryParse(token, NumberStyles.Float Or NumberStyles.AllowThousands, CultureInfo.CurrentCulture, value) Then

                If Not Double.IsNaN(value) AndAlso Not Double.IsInfinity(value) Then
                    list.Add(value)
                End If
            Else
                MessageBox.Show(Me, String.Format("Invalid numeric value: ""{0}""", token), "Parse Error", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return Nothing
            End If
        Next
        Return list
    End Function

    Private Function BuildDatasetTitle() As String
        Return "Avocado Smoothie - Barista"
    End Function

    ' Ensures the title is a valid Excel worksheet name
    Private Shared Function SanitizeWorksheetName(title As String) As String
        Dim name = If(String.IsNullOrWhiteSpace(title), "Dataset", title)

        ' Remove invalid characters: : \ / ? * [ ]
        Dim invalid As Char() = {":"c, "\"c, "/"c, "?"c, "*"c, "["c, "]"c}
        For Each ch In invalid
            name = name.Replace(ch.ToString(), String.Empty)
        Next

        name = name.Trim()

        ' Limit to 31 characters
        If name.Length > 31 Then
            name = name.Substring(0, 31)
        End If

        If String.IsNullOrWhiteSpace(name) Then
            name = "Dataset"
        End If

        Return name
    End Function

    Private Sub UpdateStatus(text As String)
        If Me.IsDisposed Then Return

        If statusStrip1 Is Nothing OrElse slblStatus Is Nothing Then Return

        If statusStrip1.InvokeRequired Then
            Try
                statusStrip1.Invoke(New Action(Sub() slblStatus.Text = text))
            Catch
                ' ignore if disposing
            End Try
        Else
            slblStatus.Text = text
        End If
    End Sub

    Private Sub SetBusy(busy As Boolean)
        Me.UseWaitCursor = busy

        If btnStart IsNot Nothing Then btnStart.Enabled = Not busy
        If btnExcelExport IsNot Nothing Then btnExcelExport.Enabled = Not busy
        If updRadius IsNot Nothing Then updRadius.Enabled = Not busy
        If updBorderCnt IsNot Nothing Then updBorderCnt.Enabled = Not busy
        If chkMiddleMedian IsNot Nothing Then chkMiddleMedian.Enabled = Not busy
        If chkAllMedian IsNot Nothing Then chkAllMedian.Enabled = Not busy
        If btnCSVExport IsNot Nothing Then btnCSVExport.Enabled = Not busy
    End Sub

    Private Shared Function SanitizeFileName(name As String) As String
        Dim fallback = "dataset"
        If String.IsNullOrWhiteSpace(name) Then Return fallback

        Dim invalid = Path.GetInvalidFileNameChars()
        Dim sb As New StringBuilder(name.Length)
        For Each ch In name.Trim()
            If Array.IndexOf(invalid, ch) >= 0 Then Continue For
            sb.Append(ch)
        Next

        Dim cleaned = sb.ToString()
        If String.IsNullOrWhiteSpace(cleaned) Then cleaned = fallback

        If cleaned.Length > 64 Then cleaned = cleaned.Substring(0, 64)
        Return cleaned
    End Function
End Class

File: Samples/C#/Winforms/AvocadoSmoothieCSSample/AvocadoSmoothieCSSample/FrmMain.cs
using AvocadoSmoothie.Barista;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using static AvocadoSmoothie.Barista.SignatureMedian;

namespace AvocadoSmoothie.Barista.Tasting
{
    public partial class FrmMain : Form
    {
        private static readonly string[] BoundaryModes = { "Symmetric", "Adaptive", "Replicate", "ZeroPad" };

        public FrmMain()
        {
            InitializeComponent();

            cbxBoundaryMode.SelectedIndex = 0; // Default selection
            UpdateParameterControls();
        }

        private void FrmMain_Load(object sender, EventArgs e)
        {
            // Ensure one method is selected by default
            if (!chkMiddleMedian.Checked && !chkAllMedian.Checked)
                chkMiddleMedian.Checked = true;

            // Remove 32K limit so both sections can be displayed for large outputs
            txtInit.MaxLength = 0;            // 0 = no limit
            txtInit.Multiline = true;         // ensure multiline (just in case)
            txtInit.WordWrap = true;

            txtRefined.MaxLength = 0;         // 0 = no limit
            txtRefined.Multiline = true;      // ensure multiline (just in case)
            txtRefined.WordWrap = true;

            UpdateStatus("Ready");
        }

        private void SmoothingMethod_CheckedChanged(object sender, EventArgs e)
        {
            UpdateParameterControls();
        }

        private void UpdateParameterControls()
        {
            // BoundaryMode : Enable AllMedian only
            cbxBoundaryMode.Enabled = chkAllMedian.Checked;
            lblBoundaryMode.Enabled = chkAllMedian.Checked;

            // BorderCount : Enable MiddleMedian only
            updBorderCnt.Enabled = chkMiddleMedian.Checked;
            lblBorderCnt.Enabled = chkMiddleMedian.Checked;
        }

        private SignatureMedian.BoundaryMode GetSelectedBoundaryMode()
        {
            switch (cbxBoundaryMode.SelectedItem?.ToString())
            {
                case "Symmetric": return SignatureMedian.BoundaryMode.Symmetric;
                case "Adaptive": return SignatureMedian.BoundaryMode.Adaptive;
                case "Replicate": return SignatureMedian.BoundaryMode.Replicate;
                case "ZeroPad": return SignatureMedian.BoundaryMode.ZeroPad;
                default: return SignatureMedian.BoundaryMode.Symmetric;
            }
        }

        private void chkMiddleMedian_CheckedChanged(object sender, EventArgs e)
        {
            SmoothingMethod_CheckedChanged(sender, e);
        }

        private void chkAllMedian_CheckedChanged(object sender, EventArgs e)
        {
            SmoothingMethod_CheckedChanged(sender, e);
        }

        private async void btnStart_Click(object sender, EventArgs e)
        {
            try
            {
                SetBusy(true);

                var data = ParseInputData(txtInit.Text);
                if (data == null || data.Count == 0)
                {
                    MessageBox.Show(this, "Please provide at least one numeric value in the Initial Data box.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    UpdateStatus("No data to process");
                    return;
                }

                int radius = DecimalToIntSafe(updRadius.Value); // R
                int borderCount = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value));
                if (radius <= 0)
                {
                    MessageBox.Show(this, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    UpdateStatus("Invalid radius");
                    return;
                }

                int windowWidth = (2 * radius) + 1; // odd window width W
                bool runAll = chkAllMedian.Checked;
                bool runMiddle = chkMiddleMedian.Checked;

                if (!runAll && !runMiddle)
                {
                    MessageBox.Show(this, "Please select at least one method: All Median and/or Middle Median.", "Selection Required", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    UpdateStatus("No method selected");
                    return;
                }

                // Clamp Middle Median border count so 2B < (2R + 1) => B <= R - 1
                int middleBorder = borderCount;
                int maxValidB = Math.Max(0, radius - 1);
                if (runMiddle && middleBorder > maxValidB)
                {
                    middleBorder = maxValidB;
                    UpdateStatus($"Middle Median: Border Count clamped to {middleBorder} (R={radius})");
                }

                UpdateStatus($"Smoothing started (R={radius}, W={windowWidth})...");
                txtRefined.Clear();

                var n = data.Count;
                var sb = new StringBuilder();

                // Run All Median if requested
                if (runAll)
                {
                    var progressAll = new Progress<int>(v =>
                    {
                        int percent = n > 0 ? Math.Min(100, Math.Max(0, (int)Math.Round((v / (double)n) * 100))) : 0;
                        UpdateStatus($"AllMedian progress: {percent}%");
                    });

                    var boundaryMode = GetSelectedBoundaryMode();

                    var allResult = await Task.Run(() =>
                        SignatureMedian.ComputeMedians(
                            input: data.ToList(),
                            useMiddle: false,
                            kernelWidth: windowWidth,  // API expects W (odd window width)
                            borderCount: borderCount,
                            progress: progressAll,
                            boundaryMode: boundaryMode));

                    sb.AppendLine("[All Median]");
                    foreach (var val in allResult)
                        sb.AppendLine(val.ToString("G17", CultureInfo.InvariantCulture));

                    if (runMiddle)
                    {
                        sb.AppendLine();
                    }
                }

                // Run Middle Median if requested
                if (runMiddle)
                {
                    var progressMiddle = new Progress<int>(v =>
                    {
                        int percent = n > 0 ? Math.Min(100, Math.Max(0, (int)Math.Round((v / (double)n) * 100))) : 0;
                        UpdateStatus($"MiddleMedian progress: {percent}%");
                    });

                    var middleResult = await Task.Run(() =>
                        SignatureMedian.ComputeMedians(
                            input: data.ToList(),
                            useMiddle: true,
                            kernelWidth: windowWidth,    // API expects W (odd window width)
                            borderCount: middleBorder,   // clamped if needed
                            progress: progressMiddle));

                    sb.AppendLine("[Middle Median]");
                    foreach (var val in middleResult)
                        sb.AppendLine(val.ToString("G17", CultureInfo.InvariantCulture));
                }

                txtRefined.Text = sb.ToString().TrimEnd();
                UpdateStatus("Smoothing completed");
            }
            catch (ArgumentOutOfRangeException ex)
            {
                MessageBox.Show(this, ex.Message, "Out of Range", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Out of range");
            }
            catch (ArgumentException ex)
            {
                MessageBox.Show(this, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Argument error");
            }
            catch (InvalidOperationException ex)
            {
                MessageBox.Show(this, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Operation error");
            }
            catch (Exception ex)
            {
                MessageBox.Show(this, ex.Message, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Error");
            }
            finally
            {
                SetBusy(false);
            }
        }

        private async void btnExcelExport_Click(object sender, EventArgs e)
        {
            CancellationTokenSource cts = null;
            try
            {
                SetBusy(true);

                var data = ParseInputData(txtInit.Text);
                if (data == null || data.Count == 0)
                {
                    MessageBox.Show(this, "Please provide at least one numeric value in the Initial Data box before exporting.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    UpdateStatus("No data to export");
                    return;
                }

                int radius = DecimalToIntSafe(updRadius.Value);   // R
                int borderCount = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value));
                if (radius <= 0)
                {
                    MessageBox.Show(this, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    UpdateStatus("Invalid radius");
                    return;
                }

                int windowWidth = (2 * radius) + 1; // W = 2R + 1
                if (chkMiddleMedian.Checked && (2 * borderCount) >= windowWidth)
                {
                    var resp = MessageBox.Show(this,
                        "For Middle Median, 2 × Border Count must be strictly less than Kernel Width (W = 2R + 1).\n\nContinue anyway?",
                        "Parameter Warning",
                        MessageBoxButtons.YesNo,
                        MessageBoxIcon.Warning);
                    if (resp == DialogResult.No)
                    {
                        UpdateStatus("Export canceled by user");
                        return;
                    }
                }

                var progress = new Progress<int>(v =>
                {
                    UpdateStatus($"Excel export: {v}%");
                });

                cts = new CancellationTokenSource();

                // Build and sanitize dataset title to satisfy Excel worksheet name constraints
                var rawTitle = BuildDatasetTitle();
                var safeTitle = SanitizeWorksheetName(rawTitle);
                if (!string.Equals(rawTitle, safeTitle, StringComparison.Ordinal))
                {
                    UpdateStatus($"Worksheet name adjusted to: {safeTitle}");
                }

                var ticket = new ExcelOrderTicket
                {
                    InitialData = data,
                    DatasetTitle = safeTitle,   // Excel worksheet name; sanitized
                    KernelRadius = radius,      // ticket expects R (service computes W internally)
                    BorderCount = borderCount,
                    Progress = progress,
                    CancellationToken = cts.Token,
                    BoundaryMode = GetSelectedBoundaryMode()
                };

                UpdateStatus($"Launching Excel export (R={radius}, W={windowWidth})...");
                await ExcelBrewService.ExcelCustomOrder(ticket);

                UpdateStatus("Excel export completed");
            }
            catch (ArgumentException ex)
            {
                MessageBox.Show(this, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Argument error");
            }
            catch (InvalidOperationException ex)
            {
                MessageBox.Show(this, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Operation error");
            }
            catch (System.Runtime.InteropServices.COMException ex)
            {
                MessageBox.Show(this, ex.Message, "Excel Interop Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Interop error");
            }
            catch (OperationCanceledException)
            {
                UpdateStatus("Export cancelled");
            }
            catch (Exception ex)
            {
                MessageBox.Show(this, ex.Message, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Error");
            }
            finally
            {
                cts?.Dispose();
                SetBusy(false);
            }
        }

        private static int DecimalToIntSafe(decimal value)
        {
            if (value > int.MaxValue) return int.MaxValue;
            if (value < int.MinValue) return int.MinValue;
            return (int)value;
        }

        private IList<double> ParseInputData(string text)
        {
            if (string.IsNullOrWhiteSpace(text))
                return new List<double>();

            var tokens = text.Split(new[] { '\r', '\n', ',', ';', '\t', ' ' }, StringSplitOptions.RemoveEmptyEntries);
            var list = new List<double>(tokens.Length);
            foreach (var token in tokens)
            {
                if (double.TryParse(token, NumberStyles.Float | NumberStyles.AllowThousands, CultureInfo.InvariantCulture, out var value) ||
                    double.TryParse(token, NumberStyles.Float | NumberStyles.AllowThousands, CultureInfo.CurrentCulture, out value))
                {
                    if (!double.IsNaN(value) && !double.IsInfinity(value))
                        list.Add(value);
                }
                else
                {
                    MessageBox.Show(this, $"Invalid numeric value: \"{token}\"", "Parse Error", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    return null;
                }
            }
            return list;
        }

        private string BuildDatasetTitle()
        {
            return $"Avocado Smoothie - Barista";
        }

        // Ensures the title is a valid Excel worksheet name
        private static string SanitizeWorksheetName(string title)
        {
            string name = string.IsNullOrWhiteSpace(title) ? "Dataset" : title;
            // Remove invalid characters: : \ / ? * [ ]
            char[] invalid = { ':', '\\', '/', '?', '*', '[', ']' };
            foreach (var ch in invalid)
                name = name.Replace(ch.ToString(), string.Empty);

            name = name.Trim();

            // Limit to 31 characters
            if (name.Length > 31)
                name = name.Substring(0, 31);

            if (string.IsNullOrWhiteSpace(name))
                name = "Dataset";

            return name;
        }

        private void UpdateStatus(string text)
        {
            if (IsDisposed) return;
            if (statusStrip1.InvokeRequired)
            {
                try { statusStrip1.Invoke(new Action(() => slblStatus.Text = text)); } catch { /* ignore if disposing */ }
            }
            else
            {
                slblStatus.Text = text;
            }
        }

        private void SetBusy(bool busy)
        {
            this.UseWaitCursor = busy;
            btnStart.Enabled = !busy;
            btnExcelExport.Enabled = !busy;
            updRadius.Enabled = !busy;
            updBorderCnt.Enabled = !busy;
            chkMiddleMedian.Enabled = !busy;
            chkAllMedian.Enabled = !busy;
            btnCSVExport.Enabled = !busy;
        }

        private async void btnCSVExport_Click(object sender, EventArgs e)
        {
            CancellationTokenSource cts = null;
            try
            {
                SetBusy(true);

                var data = ParseInputData(txtInit.Text);
                if (data == null || data.Count == 0)
                {
                    MessageBox.Show(this, "Please provide at least one numeric value in the Initial Data box before exporting.", "Input Required", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    UpdateStatus("No data to export");
                    return;
                }

                int radius = DecimalToIntSafe(updRadius.Value);   // R
                int borderCount = Math.Max(0, DecimalToIntSafe(updBorderCnt.Value));
                if (radius <= 0)
                {
                    MessageBox.Show(this, "Kernel Radius must be greater than 0.", "Invalid Radius", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    UpdateStatus("Invalid radius");
                    return;
                }

                int windowWidth = (2 * radius) + 1; // W
                if (chkMiddleMedian.Checked && (2 * borderCount) >= windowWidth)
                {
                    var resp = MessageBox.Show(this,
                        "For Middle Median, 2 × Border Count must be strictly less than Kernel Width (W = 2R + 1).\n\nContinue anyway?",
                        "Parameter Warning",
                        MessageBoxButtons.YesNo,
                        MessageBoxIcon.Warning);
                    if (resp == DialogResult.No)
                    {
                        UpdateStatus("CSV export canceled by user");
                        return;
                    }
                }

                // Ask user where to save CSV
                var title = BuildDatasetTitle();
                var sfd = new SaveFileDialog
                {
                    Title = "Save CSV",
                    Filter = "CSV files (*.csv)|*.csv|All files (*.*)|*.*",
                    FileName = SanitizeFileName(title) + ".csv",
                    AddExtension = true,
                    OverwritePrompt = true,
                    SupportMultiDottedExtensions = true
                };

                if (sfd.ShowDialog(this) != DialogResult.OK)
                {
                    UpdateStatus("CSV export canceled");
                    return;
                }

                var progress = new Progress<int>(v =>
                {
                    UpdateStatus($"CSV export: {v}%");
                });

                cts = new CancellationTokenSource();

                var ticket = new CsvOrderTicket
                {
                    InitialData = data,
                    DatasetTitle = title,
                    KernelRadius = radius,   // ticket expects R; service computes W internally
                    BorderCount = borderCount,
                    BasePath = sfd.FileName,
                    Progress = progress,
                    CancellationToken = cts.Token,
                    BoundaryMode = GetSelectedBoundaryMode()
                };

                UpdateStatus($"Launching CSV export (R={radius}, W={windowWidth})...");
                await CsvBrewService.ExportCsvAsync(ticket);

                UpdateStatus("CSV export completed");
            }
            catch (ArgumentException ex)
            {
                MessageBox.Show(this, ex.Message, "Argument Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Argument error");
            }
            catch (DirectoryNotFoundException ex)
            {
                MessageBox.Show(this, ex.Message, "Directory Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Directory error");
            }
            catch (InvalidOperationException ex)
            {
                MessageBox.Show(this, ex.Message, "Operation Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Operation error");
            }
            catch (OperationCanceledException)
            {
                UpdateStatus("CSV export canceled");
            }
            catch (Exception ex)
            {
                MessageBox.Show(this, ex.Message, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                UpdateStatus("Error");
            }
            finally
            {
                cts?.Dispose();
                SetBusy(false);
            }
        }

        private static string SanitizeFileName(string name)
        {
            var fallback = "dataset";
            if (string.IsNullOrWhiteSpace(name)) return fallback;

            var invalid = Path.GetInvalidFileNameChars();
            var sb = new StringBuilder(name.Length);
            foreach (var ch in name.Trim())
            {
                if (Array.IndexOf(invalid, ch) >= 0) continue;
                sb.Append(ch);
            }

            var cleaned = sb.ToString();
            if (string.IsNullOrWhiteSpace(cleaned)) cleaned = fallback;

            if (cleaned.Length > 64) cleaned = cleaned.Substring(0, 64);
            return cleaned;
        }
    }
}